### DISCUSSION TITLE
Keycloak 24.0.5 - two instances under one domain using seperate paths.

### DISCUSSION THREAD

**Discussion by james-bellamy** [keycloak/keycloak#33782] [Q&A]
Hello, I'm attempting to run two keycloaks under the same base domain, but separated by the paths. That is to say KC 1 answers requests to <domain>/idp and KC 2 answers requests to <domain>/idp-inactive. 

The variables in common:
```
KC_HOSTNAME_STRICT = false
KC_PROXY = edge
KC_HTTP_RELATIVE_PATH = idp (This is at build time)
```

KC 1 specific variables:
```
KC_HOSTNAME_PATH = idp
```

KC 2 specific variables:
```
KC_HOSTNAME_PATH = idp-inactive
```

`/idp/admin/products/console/` works as expected but `/idp-inactive/admin/products/console/` returns `Resource not found`

if I remove `KC_HTTP_RELATIVE_PATH` then both endpoints return `Keycloak We are Sorry.... page not found`

Is there an issue with my configuration or is this not possible?

**Comment by stianst** [comment]
Did you try accessing the instances directly (and not through the.revwrse proxy)?

What is the error in the logs?

Are you pointing both to the same DB? Are both in the same cluster (you can see in logs of they do that).

**Reply by james-bellamy** [reply]
Thank you for getting back to me so quickly, currently they're pointing at the same DB cluster but two different databases within that, the plan is for them to have their own RDS cluster. They're also in separate infinispan clusters. I don't see anything in the logs relating to the error message.

I'm going to work on accessing the directly instead of via the reverse proxy

**Reply by james-bellamy** [reply]
I've now tried accessing them directly and I still get the same behaviour with /idp working as expected and the other one returning 404

**Comment by darius-m** [comment]
I think the relative path should differ between the two - `/idp` and `/idp-inactive`, respectively, and not common; it's likely why the `/idp-inactive` resources are not served, as the paths do not match.

You may need to also set the [hostname](https://www.keycloak.org/server/hostname#_defining_specific_parts_of_the_hostname_option) accordingly to reflect the entire base path, but maybe that's only true if strict hostname checking is enabled.

**Reply by james-bellamy** [reply]
I've switched to useing /idp and /inactive so that there is no commonality between the two paths but I'm still getting the same behaviour

**Comment by james-bellamy** [comment]
Based on the hostname debug it looks like the admin console doesn't respect the `KC_HOSTNAME_PATH` value, although I'm not sure why it's showing inactive for the frontend and backend when it's set to "inactive" - I'm going to try setting the admin URL and seeing if that improve things

![Screenshot 2024-10-14 at 17 33 06](https://github.com/user-attachments/assets/afc6d854-db50-48eb-861a-3058494a7758)
![Screenshot 2024-10-14 at 17 34 09](https://github.com/user-attachments/assets/fc67cb8a-de1d-4fb0-b559-081ab1ef673b)

**Reply by darius-m** [reply]
Could you please provide how the variables are defined for each instance? For common variables it may be more clear if you would repeat them, so the configs are "complete" for each instance.

**Reply by codespearhead** [reply]
Please confirm that this issue is still reproducible in the latest stable release of Keycloak, which currently is version `26.0.0`, and which uses Hostname v2 [1], then:

Change this:

```bash
KC_PROXY=edge
KC_HOSTNAME_STRICT=false
KC_HOSTNAME_PATH=idp
```

To this:

```bash
KC_PROXY_HEADERS=xforwarded
KC_HOSTNAME_STRICT=false
KC_HOSTNAME=my.keycloak.org:123/idp
```

Also, let's reply in thread instead of creating new threads to make the conversation easier to follow.

[1] https://github.com/keycloak/keycloak/blob/10aca5552314d6e5c9226f3c6d0107861c38c693/docs/documentation/upgrading/topics/changes/changes-25_0_0.adoc#new-hostname-options

**Reply by james-bellamy** [reply]
@codespearhead until we've worked out how to perform a keycloak upgrade without noticeable downtime I'm unable to upgrade any of my environments to Keycloak 26. For the time being, I can't comment on the behavior on any version except from 24.0.5. 

`KC_HOSTNAME` seems to be a hostname v2 option but I have to use hostname v1. The equivalent appears to be `KC_HOSTNAME_URL` / `KC_HOSTNAME_ADMIN_URL` which I've set to `<domain>/idp` and `<domain>/inactive` but that doesn't seem to improve the situation. I tried moving from `proxy=edge` to `proxy_headers=xforwarded` but it produced an error relating to security certificates and that's not something I want to try and fix at the moment as according to the docs `proxy=edge` should result in the same behavior

**Reply by codespearhead** [reply]
At this point, it's a dead-end because only the latest Keycloak release receives updates and support [1][2]. Rolling upgrades are not supported [3][4].

[1] https://github.com/keycloak/keycloak/discussions/29126
[2] https://github.com/keycloak/keycloak/discussions/29126#discussioncomment-9257666
[3] https://github.com/keycloak/keycloak/discussions/26243
[4] https://github.com/keycloak/keycloak/discussions/26243#discussioncomment-8348697

**Reply by james-bellamy** [reply]
I'm aware that only the latest keycloak receives updates, we can't upgrade to the latest keycloak without causing downtime, so I'm trying to configure two parallel keycloak instances so that we can update one while using the other, and then switch once the update is complete.

**Reply by KHADIJAFD** [reply]
Hello James,
I hope you're doing well.
Could you please provide me with guidelines on how to set up and manage multiple instances of Keycloak? Any best practices or recommendations would be greatly appreciated.
